multiproof.go

```diff
@@ -274,3 +278,55 @@ func (mp MultiProof) Equal(other MultiProof) bool {
 	}
 	return mp.D.Equal(&other.D)
 }
+
+func groupPolynomialsByEvaluationPoint(fs [][]fr.Element, powersOfR []fr.Element, zs []uint8) [common.VectorLength][]fr.Element {
+	workersAggregations := make(chan [common.VectorLength][]fr.Element)
+
+	numWorkers := runtime.NumCPU()
+	batchSize := (len(fs) + numWorkers - 1) / numWorkers
+	for i := 0; i < numWorkers; i++ {
+		go func(start, end int) {
+			if end > len(fs) {
+				end = len(fs)
+			}
+			var groupedFs [common.VectorLength][]fr.Element
+			for i := start; i < end; i++ {
+				z := zs[i]
+				if len(groupedFs[z]) == 0 {
+					groupedFs[z] = make([]fr.Element, common.VectorLength)
+				}
+
+				for j := 0; j < common.VectorLength; j++ {
+					var scaledEvaluation fr.Element
+					scaledEvaluation.Mul(&powersOfR[i], &fs[i][j])
+					groupedFs[z][j].Add(&groupedFs[z][j], &scaledEvaluation)
+				}
+			}
+			workersAggregations <- groupedFs
+		}(i*batchSize, (i+1)*batchSize)
+	}
```

Comment:
We split the `fs` grouping by `zs` uniformly in (up to) `runtime.NumCPU()` goroutines.

Each go routine will send it's aggregated result per evaluation point domain in the `workersAggregations` channel.